File: bookshelf.java
import java.util.*;

// Page class implementation
class Page {
    private int pageNumber;
    private int pageLength;
    private List<String> pageData;

    public Page(int number, String data, int pageLength) {
        this.pageNumber = number;
        this.pageLength = pageLength;
        this.pageData = new ArrayList<>(Collections.nCopies(pageLength, data));
    }

    public boolean searchPage(String data) {
        return pageData.contains(data);
    }

    public int getPageNumber() {
        return pageNumber;
    }

    public List<String> getPageData() {
        return pageData;
    }
}

// Book class implementation
class Book {
    private String title;
    private String author;
    private List<Page> pages;

    public Book(String title, String author) {
        this.title = title;
        this.author = author;
        this.pages = new ArrayList<>();
    }

    public static boolean access(Book book, String author) {
        if (book.author.equals(author)) {
            return true;
        } else if (author.equals("Unknown")) {
            return false;
        } else {
            return false;
        }
    }

    public Page newPage(int number, String data, int pageLength) {
        Page page = new Page(number, data, pageLength);
        pages.add(page);
        return page;
    }

    public String getAuthor() {
        return author;
    }

    public String getTitle() {
        return title;
    }

    public boolean pushPage(Page page) {
        pages.add(page);
        return true;
    }

    public boolean popPage() {
        if (!pages.isEmpty()) {
            pages.remove(pages.size() - 1);
            return true;
        } else {
            System.err.println("No pages to remove.");
            return false;
        }
    }

    public boolean insertPage(Page page, int position) {
        if (position < 1 || position > pages.size() + 1) {
            System.err.println("Position out of bounds.");
            return false;
        }
        pages.add(position - 1, page);
        return true;
    }

    public List<String> viewPages(String titleFilter) {
        List<String> result = new ArrayList<>();
        for (Page page : pages) {
            if (titleFilter.isEmpty() || title.equals(titleFilter)) {
                result.add(page.getPageNumber() + ": " + (page.getPageData().isEmpty() ? "" : page.getPageData().get(0)));
            }
        }
        return result;
    }
}

// Shelf class implementation
class Shelf {
    private List<Book> books;

    public Shelf() {
        this.books = new ArrayList<>();
    }

    public boolean pushBook(Book book) {
        books.add(book);
        return true;
    }

    public boolean popBook() {
        if (!books.isEmpty()) {
            books.remove(books.size() - 1);
            return true;
        } else {
            System.err.println("No books to remove.");
            return false;
        }
    }

    public boolean insertBook(Book book, int position) {
        if (position < 1 || position > books.size() + 1) {
            System.err.println("Position out of bounds.");
            return false;
        }
        books.add(position - 1, book);
        return true;
    }

    public List<String> viewBooks(String titleFilter) {
        List<String> result = new ArrayList<>();
        for (Book book : books) {
            if (titleFilter.isEmpty() || book.getTitle().equals(titleFilter)) {
                result.add(book.getTitle());
            }
        }
        return result;
    }

    public List<String> searchBookshelf(String term) {
        List<String> results = new ArrayList<>();
        for (Book book : books) {
            if (book.getTitle().contains(term)) {
                results.add("Title: " + book.getTitle());
            }
            for (String page : book.viewPages("")) {
                if (page.contains(term)) {
                    results.add("Page Data in book: " + book.getTitle());
                }
            }
        }
        return results;
    }

    public void sortBookTitle() {
        books.sort(Comparator.comparing(Book::getTitle));
    }

    public void sortBookAuthor() {
        books.sort(Comparator.comparing(Book::getAuthor));
    }
}

File: bookshelf.rs
use std::cmp::Ordering;

// Page struct implementation
pub struct Page {
    page_number: i32,
    page_length: i32,
    page_data: Vec<String>,
}

impl Page {
    pub fn new(number: i32, data: &str, page_length: i32) -> Self {
        Page {
            page_number: number,
            page_length,
            page_data: vec![data.to_string(); page_length as usize],
        }
    }

    pub fn search_page(&self, data: &str) -> bool {
        self.page_data.contains(&data.to_string())
    }

    pub fn get_page_number(&self) -> i32 {
        self.page_number
    }

    pub fn get_page_data(&self) -> &Vec<String> {
        &self.page_data
    }
}

// Book struct implementation
pub struct Book {
    title: String,
    author: String,
    pages: Vec<Page>,
}

impl Book {
    pub fn new(title: &str, author: &str) -> Self {
        Book {
            title: title.to_string(),
            author: author.to_string(),
            pages: Vec::new(),
        }
    }

    pub fn access(&self, author: &str) -> bool {
        if self.author == author {
            true
        } else if author == "Unknown" {
            false
        } else {
            false
        }
    }

    pub fn new_page(&mut self, number: i32, data: &str, page_length: i32) -> Page {
        let page = Page::new(number, data, page_length);
        self.pages.push(page.clone());
        page
    }

    pub fn get_author(&self) -> &str {
        &self.author
    }

    pub fn get_title(&self) -> &str {
        &self.title
    }

    pub fn push_page(&mut self, page: Page) -> bool {
        self.pages.push(page);
        true
    }

    pub fn pop_page(&mut self) -> bool {
        if !self.pages.is_empty() {
            self.pages.pop();
            true
        } else {
            eprintln!("No pages to remove.");
            false
        }
    }

    pub fn insert_page(&mut self, page: Page, position: usize) -> bool {
        if position < 1 || position > self.pages.len() + 1 {
            eprintln!("Position out of bounds.");
            false
        } else {
            self.pages.insert(position - 1, page);
            true
        }
    }

    pub fn view_pages(&self, title_filter: &str) -> Vec<String> {
        self.pages
            .iter()
            .filter(|_| title_filter.is_empty() || self.title == title_filter)
            .map(|page| {
                format!(
                    "{}: {}",
                    page.get_page_number(),
                    page.get_page_data().first().unwrap_or(&String::new())
                )
            })
            .collect()
    }
}

// Shelf struct implementation
pub struct Shelf {
    books: Vec<Book>,
}

impl Shelf {
    pub fn new() -> Self {
        Shelf { books: Vec::new() }
    }

    pub fn push_book(&mut self, book: Book) -> bool {
        self.books.push(book);
        true
    }

    pub fn pop_book(&mut self) -> bool {
        if !self.books.is_empty() {
            self.books.pop();
            true
        } else {
            eprintln!("No books to remove.");
            false
        }
    }

    pub fn insert_book(&mut self, book: Book, position: usize) -> bool {
        if position < 1 || position > self.books.len() + 1 {
            eprintln!("Position out of bounds.");
            false
        } else {
            self.books.insert(position - 1, book);
            true
        }
    }

    pub fn view_books(&self, title_filter: &str) -> Vec<String> {
        self.books
            .iter()
            .filter(|book| title_filter.is_empty() || book.get_title() == title_filter)
            .map(|book| book.get_title().to_string())
            .collect()
    }

    pub fn search_bookshelf(&self, term: &str) -> Vec<String> {
        let mut results = Vec::new();
        for book in &self.books {
            if book.get_title().contains(term) {
                results.push(format!("Title: {}", book.get_title()));
            }
            for page in book.view_pages("") {
                if page.contains(term) {
                    results.push(format!("Page Data in book: {}", book.get_title()));
                }
            }
        }
        results
    }

    pub fn sort_book_title(&mut self) {
        self.books.sort_by(|a, b| a.get_title().cmp(b.get_title()));
    }

    pub fn sort_book_author(&mut self) {
        self.books.sort_by(|a, b| a.get_author().cmp(b.get_author()));
    }
}

